File: coinChange.go
// problem: https: // leetcode.com/problems/coin-change/
// Runtime: 4 ms, faster than 99.29% of Go online submissions for Coin Change.
// Memory Usage: 6 MB, less than 85.41% of Go online submissions for Coin Change.

package algorithms

import "math"

func min(a int, b int) int {
	if a <= b {
		return a
	}
	return b
}

func coinChange(coins []int, amount int) int {
	dp := make([]int, amount+1)
	dp[0] = 0
	for i := 1; i <= amount; i++ {
		dp[i] = math.MaxInt32
	}
	for i, _ := range dp {
		for j, _ := range coins {
			if i-coins[j] < 0 {
				continue
			}
			dp[i] = min(dp[i], dp[i-coins[j]]+1)
		}
	}
	if dp[amount] == math.MaxInt32 {
		return -1
	}
	return dp[amount]
}

File: coinChange.py
# problem: https: // leetcode.com/problems/coin-change/
# Runtime: 1296 ms, faster than 60.21 % of Python3 online submissions for Coin Change.
# Memory Usage: 14.1 MB, less than 97.76 % of Python3 online submissions for Coin Change.

from typing import List


class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [float("inf")] * (amount + 1)
        dp[0] = 0
        for i in range(len(dp)):
            for coin in coins:
                if i - coin < 0:
                    continue
                dp[i] = min(dp[i], dp[i-coin] + 1)
        return dp[-1] if dp[-1] != float("inf") else -1
